File: inclass/conjugateGradient.m
function [x, objectiveValue, g, xHistory, objectiveHistory] = mymin5(initialX)
   Q = [2, 1; 1, 2];
   b = [10; 10];

   % If the initial values will overflow, then just take the log of them.
   if (initialX' * initialX == Inf || exp(initialX(1)) == Inf || exp(initialX(2)) == Inf)
      initialX = [log(initialX(1)); log(initialX(2))];
   end;

   x = initialX;
   xHistory = initialX;                        % xHistory will store all the x's we visit

   objectiveValue = .5 * x' * Q * x - b' * x + exp(x(1) / 10) + exp(2 * x(2) / 10);
   objectiveHistory = objectiveValue;                      % objectiveHistory array stores all fvals we see
   oldObjectiveValue = Inf;  % Since we are just starting, best objective found in past step is infinity

   while abs(objectiveValue - oldObjectiveValue) > 1e-12
      H = hessian(x, Q);              % Compute Hessian to use for next n = 2 steps
                                      % (This is the 'Q' of the quadratic approximation)
      g_center = gradient(x, Q, b)';  % Compute gradient at the center of the quadratic approximation
      b_lin = H * x - g_center;       % This finds the 'b' term of the quadratic approximation

      for j = 1:2                     % Loop through n = 2 steps
         g = H * x - b_lin;           % Compute gradient from quadratic approximation
         if j == 1
               d = -g;                % First step should be a gradient step
         else
               gamma = (g' * H * d) / (d' * H * d);
               d = -(g - gamma * d);
         end;

         % If gradient miniscule, avoid dividing by something close to zero
         if norm(g) > 1e-9
            % Choose step size to minimize quadratic approximation:
            % alpha = -g' * d / (d' * H * d);

            % Uncomment below to choose step size using Newton's method
            alpha = linesearch(x, d, Q, b);
         else
            alpha = 0;
         end;

         x = x + alpha * d;  % Update x
         oldObjectiveValue = objectiveValue;
         objectiveValue = .5 * x' * Q * x - b' * x + exp(x(1) / 10) + exp(2 * x(2) / 10);
         xHistory = [xHistory, x];  % Log this x in array to keep track of our path
         objectiveHistory = [objectiveHistory, objectiveValue];  % Log this objectiveValue in array to keep track of fvals seen
      end;
   end;

   g = gradient(x, Q, b)';

function alpha = linesearch(x, d, Q, b)
   % This is a Newton's method line search
   alpha = 0;
   for j = 1:100
      alpha = alpha + -gradient(x + alpha * d, Q, b) * d  /  (d' * hessian(x + alpha * d, Q) * d);
   end;

function grad = gradient(x, Q, b)
   grad = x' * Q  - b' + (1 / 10) * [exp(x(1) / 10) , 2 * exp(2 * x(2) / 10) ];

function H = hessian(x, Q)
   H = Q + 1 / (100) * [exp(x(1) / 10), 0; 0, 4 * exp(2 * x(2) / 10)];

File: assignments/07/conjugateGradient.py
import math
import numpy

CONVERGENCE_DIFF = 1e-15
GRADIENT_CUTOFF = 1e-9

# All matric work, input, and output will be done with numpy.matrix.
# Returns: [x, objectiveValue, g, xHistory, objectiveHistory]
def conjugateGradient(initialX, objective, gradient, hessian):
    # If the initial values will overflow, then just take the log of them.
    try:
        _ = (initialX.T * initialX)[0, 0]
        for row in range(initialX.shape[0]):
            for col in range(initialX.shape[1]):
                _ += math.exp(initialX[row, col])
    except OverflowError:
        for row in range(initialX.shape[0]):
            for col in range(initialX.shape[1]):
                initialX[row, col] = math.log(initialX[row, col])

    x = initialX

    oldObjectiveValue = None
    objectiveValue = objective(x)

    # Keep track of some historical values.
    xHistory = [initialX.copy()]
    objectiveHistory = [objectiveValue]

    while (oldObjectiveValue == None or abs(objectiveValue - oldObjectiveValue) > CONVERGENCE_DIFF):
        # Compute Hessian to use for next n = 2 steps
        # (This is the "Q" of the quadratic approximation)
        H = hessian(x)

        # Compute gradient at the center of the quadratic approximation
        g_center = gradient(x).T

        # This finds the "b" term of the quadratic approximation
        b_lin = H * x - g_center

        # Loop through n = 2 steps
        for n in range(2):
            # Compute gradient from quadratic approximation
            g = H * x - b_lin

            if n == 0:
                # First step should be a gradient step
                d = -g
            else:
                gamma = (g.T * H * d)[0, 0] / (d.T * H * d)[0, 0]
                d = -(g - gamma * d)

            # If gradient miniscule, avoid dividing by something close to zero
            if (numpy.linalg.norm(g) > GRADIENT_CUTOFF):
                # Choose step size to minimize quadratic approximation:
                # alpha = -g.T * d / (d.T * H * d)

                # Uncomment below to choose step size using Newton's method
                alpha = linesearch(x, d, gradient, hessian)
            else:
                alpha = 0

            # Update x
            x += alpha * d

            oldObjectiveValue = objectiveValue
            objectiveValue = objective(x)

            # Log histories
            xHistory.append(x.copy())
            objectiveHistory.append(objectiveValue)

    g = gradient(x).T

    return [
        x,
        objectiveValue,
        g,
        xHistory,
        objectiveHistory
    ]

def l2Norm(vector):
    sum = 0
    for value in vector:
        sum += value * value

    return math.sqrt(sum)

def linesearch(x, d, gradient, hessian):
    # This is a Newton's method line search
    alpha = 0
    for i in range(100):
        alpha += -(gradient(x + alpha * d) * d)[0, 0] / (d.T * hessian(x + alpha * d) * d)[0, 0]

    return alpha
